python/ml_croissant/_src/operation_graph/base_operation.py

```diff
@@ -0,0 +1,29 @@
+"""Base operation module."""
+
+import dataclasses
+
+from ml_croissant._src.nodes import Node
+
+
+@dataclasses.dataclass(frozen=True, repr=False)
+class Operation:
+    """Generic base class to define an operation.
+
+    `@dataclass(frozen=True)` allows having a hashable operation for NetworkX to use
+    operations as nodes of  graphs.
+
+    `@dataclass(repr=False)` allows having a readable stringified `str(operation)`.
+
+    Args:
+        node: The node attached to the operation for the context.
+        output: The result of the operation when it is executed (as returned by
+            __call__).
+    """
+
+    node: Node
```

Comment:
This `node` is the input of the `__call__`. The output type is now not defined. I think that it would make defining the graphs easier if we would have type hints on the output. That would make it possible to automatically check if an Operation can follow another Operation (I assume here that the output of `__call__` of an operation, might be needed for the next operation. I'm not sure exactly how this is implemented, so I might misunderstand some parts). 
It could be implemented using generics. Maybe this is over-engineering, but I always like it.